src/main/java/io/odpf/firehose/sink/clickhouse/QueryTemplate.java

```diff
@@ -0,0 +1,118 @@
+package io.odpf.firehose.sink.clickhouse;
+
+import com.samskivert.mustache.Escapers;
+import com.samskivert.mustache.Mustache;
+import com.samskivert.mustache.Template;
+import io.odpf.firehose.config.ClickhouseSinkConfig;
+import io.odpf.firehose.message.Message;
+import io.odpf.firehose.proto.ProtoToFieldMapper;
+
+import java.util.List;
+import java.util.HashMap;
+import java.util.Properties;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Query template for clickhouse.
+ */
+public class QueryTemplate {
+    private static final String INSERT_QUERY = "INSERT INTO {{table}} ( {{insertColumns}} ) values ( {{insertValues}} ) ";
+    private static final String VALUES_SEPERATOR = " ),( "; // used for multiple values in a single query.
+    private Template template;
+    private ProtoToFieldMapper protoToFieldMapper;
+    private List<String> insertColumns;
+    private HashMap<String, Object> scopes;
+    private String kafkaRecordParserMode;
+
+    public QueryTemplate(ClickhouseSinkConfig clickhouseSinkConfig, ProtoToFieldMapper protoToFieldMapper) {
+        this.protoToFieldMapper = protoToFieldMapper;
+        this.insertColumns = new ArrayList<>();
+        this.scopes = new HashMap<>();
+        this.kafkaRecordParserMode = clickhouseSinkConfig.getKafkaRecordParserMode();
+        initialize(clickhouseSinkConfig);
+        buildQuery();
+    }
+
+    private void buildQuery() {
+        String query = INSERT_QUERY;
+        template = Mustache.compiler().withEscaper(Escapers.simple()).compile(query);
+    }
+
+    private void initialize(ClickhouseSinkConfig clickhouseSinkConfig) {
```

Comment:
Hm. We should call this from factory.init() method IMO and not from constructor.